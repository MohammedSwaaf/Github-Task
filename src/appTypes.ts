
export interface ISreachView {
    searchType: GitHubEntities,
    hasMore: boolean,
    loadMore: () => void;
}
export type GitHubEntities = 'repositories' | 'users'
export type Filter = {
    searchTerm: string,
    entityType: GitHubEntities,
    page: number,
    per_page: number,
}
export interface ISearchViewRepo {

    stargazers_count: string
    forks: string
    visibility: string
    watchers: string
    name: string;
    owner: {
        login: string;
        type: string;
        avatar_url: string;
    }
}
export interface ISearchViewUser {
    avatar_url: string;
    login: string;
    html_url: string;
    id: string;
    score: string;
    type: string;
}
export type GitHubReducerState = Filter & { allData: ISearchViewRepo[] & ISearchViewUser[] }
