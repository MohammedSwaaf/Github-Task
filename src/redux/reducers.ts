import { combineReducers } from 'redux';
import gitHub from './gitHubReducer/reducer';

const rootReducer = combineReducers({
    gitHubReducer:gitHub
});
export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;