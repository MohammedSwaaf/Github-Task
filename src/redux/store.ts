import { legacy_createStore as createStore, applyMiddleware} from 'redux'
import { persistStore, persistReducer } from 'redux-persist';
import thunk from 'redux-thunk';
import { composeWithDevTools } from '@redux-devtools/extension';
import storage from 'redux-persist/lib/storage';
import { combineReducers } from 'redux';
import gitHub from './gitHubReducer/reducer';

const persistConfig = {
    key: 'root',
    storage,
};

const rootReducer = combineReducers({
    gitHubReducer:gitHub
  // Add your reducers here
});
const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(persistedReducer, composeWithDevTools(applyMiddleware((thunk))));
export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>
//store structure types