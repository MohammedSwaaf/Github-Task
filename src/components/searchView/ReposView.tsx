import { useSelector } from 'react-redux';
import { truncString } from '../../utils/stringTrunc'
import { RootState } from '../../redux/store';
import { GitHubReducerState, ISearchViewRepo } from '../../appTypes';



export default function ReposView() {
    const { allData } = useSelector<RootState>(state => state.gitHubReducer) as GitHubReducerState;

    return (
        <>
            <div className="row">
                {Array.isArray(allData) && allData.map((repo: ISearchViewRepo, key: number) => {
                    return <div className="col-md-4" key={key}>
                        <div className="card">
                            <div className="card-header">
                                <h2>#{key + 1} {repo?.name ? truncString(repo?.name) : 'NA'} (Repo)</h2>
                                <div className="image-container">
                                    <img width={360} className='userImage' src={repo?.owner?.avatar_url ?? 'NA'} alt="user_photo" />
                                </div>

                            </div>
                            <div className="card-body">
                                <li>Author: <span>{repo?.owner?.login ?? 'NA'}</span></li>
                                <li>Type: <span>{repo?.owner?.type ?? 'NA'}</span></li>
                                <li>Stars: <span>{repo?.stargazers_count ?? 'NA'}</span></li>
                                <li>Forks: <span>{repo?.forks ?? 'NA'}</span></li>
                                <li>Visibility: <span>{repo?.visibility ?? 'NA'}</span></li>
                                <li>Watchers: <span>{repo?.watchers ?? 'NA'}</span></li>
                            </div>
                        </div>
                    </div>
                })}
            </div>
        </>
    )
}
