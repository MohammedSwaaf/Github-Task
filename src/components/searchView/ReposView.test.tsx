import Enzyme,{ ShallowWrapper, shallow, configure } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import toJson from 'enzyme-to-json';
import ReposView from './ReposView';

import { ISearchViewRepo } from '../../appTypes';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';

Enzyme.configure({ adapter: new Adapter() });

describe('UsersView Component', () => {
    let wrapper: ShallowWrapper;

    beforeEach(() => {

        wrapper = shallow(
            <Provider store={store}>
                <ReposView />
            </Provider>
        );
    });

    it('renders without crashing', () => {
        expect(wrapper.exists()).toBe(true);
    });

    it('renders correct number of user cards', () => {
        const repos: ISearchViewRepo[] = [
            
            {
                stargazers_count: 'user1', forks: '71', visibility: '1', watchers: '61', name: 'SwaafRepo1', owner: {
                    login: 'string',
                    type: 'string',
                    avatar_url: 'string'
                }
            },
            {
                stargazers_count: 'user2', forks: '72', visibility: '2', watchers: '62', name: 'SwaafRepo2', owner: {
                    login: 'string',
                    type: 'string',
                    avatar_url: 'string'
                }
            },
            {
                stargazers_count: 'user3', forks: '73', visibility: '3', watchers: '63', name: 'SwaafRepo3', owner: {
                    login: 'string',
                    type: 'string',
                    avatar_url: 'string'
                }
            },
        ];
        wrapper.setProps({ repos });
    });


    it('matches snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});