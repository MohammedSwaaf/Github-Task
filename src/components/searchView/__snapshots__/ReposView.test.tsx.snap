// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersView Component matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ReposView />
</ContextProvider>
`;
